### DISCUSSION TITLE
Resources for monitoring with Datadog

### DISCUSSION THREAD

**Discussion by sa-spag** [fluxcd/flux2#2054] [Proposals]
👋 Hello,

Flux v2 already comes with resources to monitor its components with Prometheus ([see related documentation](https://fluxcd.io/docs/guides/monitoring/)), but not for [Datadog](https://www.datadoghq.com/). It would be great for the community to share curated Datadog dashboards, monitors, and metrics metadata.

We could create a datadog-integration repository in https://github.com/fluxcd-community for sharing documentation on how to monitor Flux v2 with Datadog, Kubernetes manifests patches to integrate with Datadog agents, Terraform configuration or raw JSON to provision Datadog resources.

**Comment by sa-spag** [comment]
Here is a Kustomization that patches Flux's components so they can be discovered by Datadog agents. The goal is to [label Deployments and associated pods with Datadog unified service tags](https://docs.datadoghq.com/getting_started/tagging/unified_service_tagging/?tab=kubernetes#configuration-1) and [annotate pods to configure the OpenMetrics integration](https://docs.datadoghq.com/agent/kubernetes/prometheus/):

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
- gotk-components.yaml
patches:
# Annotate Flux components for Datadog
- patch: |-
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: flux-component
    spec:
      template:
        metadata:
          annotations:
            ad.datadoghq.com/manager.check_names: |
              ["openmetrics"]
            ad.datadoghq.com/manager.init_configs: |
              [{}]
            ad.datadoghq.com/manager.instances: |
              [
                {
                  "prometheus_url": "http://%%host%%:8080/metrics",
                  "namespace": "flux",
                  "send_distribution_buckets": true,
                  "metrics": [
                    "*"
                  ]
                }
              ]
  target:
    kind: Deployment
    labelSelector: app.kubernetes.io/part-of=flux
# Add Datadog unified service tags to all Flux components
patchesStrategicMerge:
- |-
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: helm-controller
      namespace: flux-system
      labels:
        tags.datadoghq.com/source: "helm-controller"
        tags.datadoghq.com/service: "helm-controller"
    spec:
      template:
        metadata:
          labels:
            tags.datadoghq.com/source: "helm-controller"
            tags.datadoghq.com/service: "helm-controller"
- |-
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: image-automation-controller
      namespace: flux-system
      labels:
        tags.datadoghq.com/source: "image-automation-controller"
        tags.datadoghq.com/service: "image-automation-controller"
    spec:
      template:
        metadata:
          labels:
            tags.datadoghq.com/source: "image-automation-controller"
            tags.datadoghq.com/service: "image-automation-controller"
- |-
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: image-reflector-controller
      namespace: flux-system
      labels:
        tags.datadoghq.com/source: "image-reflector-controller"
        tags.datadoghq.com/service: "image-reflector-controller"
    spec:
      template:
        metadata:
          labels:
            tags.datadoghq.com/source: "image-reflector-controller"
            tags.datadoghq.com/service: "image-reflector-controller"
- |-
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: image-reflector-controller
      namespace: flux-system
      labels:
        tags.datadoghq.com/source: "image-reflector-controller"
        tags.datadoghq.com/service: "image-reflector-controller"
    spec:
      template:
        metadata:
          labels:
            tags.datadoghq.com/source: "image-reflector-controller"
            tags.datadoghq.com/service: "image-reflector-controller"
- |-
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: kustomize-controller
      namespace: flux-system
      labels:
        tags.datadoghq.com/source: "kustomize-controller"
        tags.datadoghq.com/service: "kustomize-controller"
    spec:
      template:
        metadata:
          labels:
            tags.datadoghq.com/source: "kustomize-controller"
            tags.datadoghq.com/service: "kustomize-controller"
- |-
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: notification-controller
      namespace: flux-system
      labels:
        tags.datadoghq.com/source: "notification-controller"
        tags.datadoghq.com/service: "notification-controller"
    spec:
      template:
        metadata:
          labels:
            tags.datadoghq.com/source: "notification-controller"
            tags.datadoghq.com/service: "notification-controller"
- |-
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: source-controller
      namespace: flux-system
      labels:
        tags.datadoghq.com/source: "source-controller"
        tags.datadoghq.com/service: "source-controller"
    spec:
      template:
        metadata:
          labels:
            tags.datadoghq.com/source: "source-controller"
            tags.datadoghq.com/service: "source-controller"
replacements:
# Add `tags.datadoghq.com/version` label to all Flux components
- source:
    kind: Namespace
    name: flux-system
    fieldPath: metadata.labels.[app.kubernetes.io/version]
  targets:
  - select:
      kind: Deployment
      labelSelector: app.kubernetes.io/part-of=flux
    fieldPaths:
    - metadata.labels.[tags.datadoghq.com/version]
    - spec.template.metadata.labels.[tags.datadoghq.com/version]
    options:
      create: true
```

I find the whole `patchesStrategicMerge` thingy inelegant but I am not aware of alternative Kustomize features.

**Reply by juozasget** [reply]
Worked like a charm! Thank you very much for sharing @sa-spag !

Perhaps you also have some pre-built flux dashboards? If not - I'll share what I can come up with :)

**Reply by sa-spag** [reply]
Glad to hear that was useful for you! 😄

I have created this simple dashboard:
<details>
<summary>Dashboard JSON</summary>

```
{"title":"Flux","description":"","widgets":[{"id":5399722969196778,"definition":{"title":"Reconciliation status","title_size":"16","title_align":"left","type":"query_table","requests":[{"formulas":[{"alias":"Not ready","conditional_formats":[{"palette":"white_on_red","comparator":">=","value":1},{"palette":"white_on_green","comparator":"<","value":1}],"limit":{"count":500,"order":"desc"},"cell_display_mode":"number","formula":"query1"},{"alias":"Suspended","conditional_formats":[{"palette":"black_on_light_yellow","comparator":">=","value":1},{"palette":"white_on_green","comparator":"<","value":1}],"cell_display_mode":"number","formula":"query2"}],"response_format":"scalar","queries":[{"query":"max:flux.gotk_reconcile_condition{type:ready,status:false,$namespace,$name,$kind} by {kind,name,namespace}","data_source":"metrics","name":"query1","aggregator":"last"},{"query":"max:flux.gotk_suspend_status{$namespace,$name,$kind} by {kind,name,namespace}","data_source":"metrics","name":"query2","aggregator":"last"}]}],"has_search_bar":"always"},"layout":{"x":0,"y":0,"width":12,"height":4}},{"id":7997697863000238,"definition":{"title":"Workers","title_size":"16","title_align":"left","show_legend":true,"legend_layout":"vertical","legend_columns":["avg","min","max","value","sum"],"type":"timeseries","requests":[{"formulas":[{"alias":"Free workers","formula":"query2 - query1"}],"response_format":"timeseries","queries":[{"query":"max:flux.controller_runtime_max_concurrent_reconciles{*} by {pod_name}","data_source":"metrics","name":"query2"},{"query":"max:flux.controller_runtime_active_workers{*} by {pod_name}","data_source":"metrics","name":"query1"}],"style":{"palette":"dog_classic","line_type":"solid","line_width":"normal"},"display_type":"line"}],"yaxis":{"include_zero":true,"scale":"linear","label":"","min":"auto","max":"auto"},"markers":[]},"layout":{"x":0,"y":4,"width":12,"height":3}},{"id":6991964751521450,"definition":{"title":"Controller reconciliation time P90","title_size":"16","title_align":"left","show_legend":true,"legend_layout":"horizontal","legend_columns":["avg","min","max","value","sum"],"type":"timeseries","requests":[{"formulas":[{"formula":"query1"}],"response_format":"timeseries","queries":[{"query":"p90:flux.controller_runtime_reconcile_time_seconds{*} by {controller}","data_source":"metrics","name":"query1"}],"style":{"palette":"dog_classic","line_type":"solid","line_width":"normal"},"display_type":"line"}],"yaxis":{"include_zero":true,"scale":"linear","label":"","min":"auto","max":"auto"},"markers":[]},"layout":{"x":0,"y":7,"width":12,"height":3}},{"id":8043513538880860,"definition":{"title":"Events","type":"group","show_title":true,"layout_type":"ordered","widgets":[{"id":878940700469874,"definition":{"title":"Source","title_size":"16","title_align":"left","type":"event_stream","query":"sources:kubernetes tags:source_component:source-controller $namespace $name $kind","tags_execution":"and","event_size":"l"},"layout":{"x":0,"y":0,"width":12,"height":3}},{"id":7273588799650642,"definition":{"title":"Kustomization","title_size":"16","title_align":"left","type":"event_stream","query":"sources:kubernetes tags:source_component:kustomize-controller $namespace $name $kind","tags_execution":"and","event_size":"l"},"layout":{"x":0,"y":3,"width":12,"height":3}},{"id":5593602003661506,"definition":{"title":"HelmRelease","title_size":"16","title_align":"left","type":"event_stream","query":"sources:kubernetes tags:source_component:helm-controller $namespace $name $kind","tags_execution":"and","event_size":"l"},"layout":{"x":0,"y":6,"width":12,"height":3}}]},"layout":{"x":0,"y":10,"width":12,"height":10,"is_column_break":true}}],"template_variables":[{"name":"namespace","default":"*","prefix":"namespace","available_values":[]},{"name":"name","default":"*","prefix":"name","available_values":[]},{"name":"kind","default":"*","prefix":"kind","available_values":["helmrelease","helmchart","helmrepository","gitrepository","kustomization","imagepolicy","imagerepository","imageupdateautomation"]}],"layout_type":"ordered","is_read_only":false,"notify_list":[],"reflow_type":"fixed","id":"bzh-gej-34v"}
```
</details>

Note that we do not collect all metrics in order to not over consume our custom metrics budget, so depending on your setup you can show much more information.

**Reply by juozasget** [reply]
Super! Really appreciate you sharing this!

**Comment by alexandros-genies** [comment]
Thank you for this! going to try it out. Just a quick question:

Should the namespace be set to flux-system rather than flux?
```apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
- gotk-components.yaml
patches:
# Annotate Flux components for Datadog
- patch: |-
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: flux-component
    spec:
      template:
        metadata:
          annotations:
            ad.datadoghq.com/manager.check_names: |
              ["openmetrics"]
            ad.datadoghq.com/manager.init_configs: |
              [{}]
            ad.datadoghq.com/manager.instances: |
              [
                {
                  "prometheus_url": "http://%%host%%:8080/metrics",
                  "namespace": "flux",
                  "send_distribution_buckets": true,
                  "metrics": [
                    "*"
                  ]
                }
              ]```

**Reply by gaffneyd4** [reply]
The namespace you are referring to is pertaining to the openmetrics check that datadog is registering to ingest metrics from flux.
You can read more about that configuration option [here](https://docs.datadoghq.com/integrations/openmetrics/#configuration) and [here](https://github.com/DataDog/integrations-core/blob/59ae7eb02498a902b1f0fe80abbc4dd48d8d05cc/openmetrics/datadog_checks/openmetrics/data/conf.yaml.example#L54-L57).

TLDR; that value can be whatever you want to use to organize the metrics once they are ingested into the DataDog platform.

**Comment by sa-spag** [comment]
In case it went unnoticed, there's a community developed integration for Flux now: see https://github.com/fluxcd/flux2/discussions/3157! 🎉
I would suggest any Datadog related contribution to integrate with the integration from now on.

**Comment by cep21** [comment]
For people that find this ticket, you may want to use the `fluxcd` integration, not the `openmetrics` integration above since the `openmetrics` will bill as [custom metrics](https://docs.datadoghq.com/account_management/billing/custom_metrics/).  I documented this at https://github.com/fluxcd/flux2/discussions/3157
